design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,216 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+1. The `Specified-Device` does not promise to provide sufficient hooks for me to test my internal implementation of my state machine. The interfaces are different. The device manufactures did not know that I intended to create a state machine simulator, and their interface is different to the one that I want for my state machine.
+1. The `Specified-Device` interface includes conversions that are not fundamental to my state-machine and my state-machine does not rely on those conversions. For example, the `Specified-Device` returns a temperature as `32` degrees as `0x020`.
+    1. I want to be able to verify that the device is processing say a temperature limit set `L1` to the correct temperature independently of what the `State-Machine-Device` then does with it. I want failures to easily indicate where problems are in the code base (see premise above). Are they related to the state-machine or the parsing of inputs?
+    1. I want to have readable tests for my `State-Machine-Device` a set/return `0x020` is not incredibly readable at the unit test level (although it is needed at the system-test level)
+
+    
+I would then say as part of my proposal that the low-level design (as created by the Plankton team) of stateful devices might better look like this.
+
+```python
+
+class LinkumT95()
+
+   def __init__(self, linkum_t95_machine):
+       if not linkum_t95_machine:
+           self._linkum_t95_machine = LinkumT95Machine()
+       else:
+           self._linkum_t95_machine = linkum_t95_machine
+    
+   def getStatus():
+       temp = self._linkum_t95_machine.getTemp() # Get temp as float
+       # convert this into a hex 
+       # convert this into a hex character array
+       return hex_char_arry
+   
+```
+This will give a better ability to perform the granular testing at the level that we want. 
+
+ 
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
```

Comment:
Talked about this with Owen last week, but forgot to leave a comment here: The delta t would be controlled by a multiplier (which defaults to 1) rather than a fixed override as shown here. Not sure if we want to update the document or just consider that an implementation detail.